library/Fission/Web.hs

```diff
@@ -57,6 +60,7 @@ app handlerNT auth appHost = do
 -- | Web handlers for the 'API'
 server ::
   ( MonadDB               m
+  , MonadDBQuery UserCid m
```

Comment:
@expede Got a version of this working just far enough to see what the type signature of this approach would be.

I gotta say I'm not loving that at the top we'd have to define a type constraint for every query type.

```haskell
MonadDB m
=> MonadDBQuery UserCid m
=> MonadDBQuery User m
=> MonadDBQuery App m
=> MonadDBQuery Domain m
```

I imagine theres a better way or perhaps a technique to avoid this that I dont know. WDYT?

Reply:
> I'm not loving that at the top we'd have to define a type constraint for every query type.

I mean, being explicit is the entire point here. If we just want unconstrained DB access, then it's pretty straightforward — don't use these and leave it at `MonadDB`. I'm okay with that, as per our call yesterday.

Breaking them out is a tradeoff. It's really not that much boilerplate, and you get a lot more resource control. IMO tracing what a function can touch is more of a thing for mutations than queries, so the choice for these is totally up to you. Breaking queries out will make mocking out tests easier (because now we have granular access)

# Alias

If you want to alias several of these together, that's totally doable though I say discourage it in this case:

```haskell
type MonadQueryAllModels m = (MonadDBQuery User m, MonadDBQUery Domain m, ...)
```

The problem here is that it temps people to use widest constraint out of laziness/convenience, this negating most of the value.

# Default Implementation

To get around having to _define_ these, you can use a default implementation:

```haskell
class MonadDBQuery model m where
  getBy :: SqlExpr something something -> Transaction m [Entity model]
  getBy selector = select <| from \model ->
    where_ (selector model)
    return model
```

Now you get all of the models implemented out of the box with the same version, but you can still overwrite specific cases if you need to (e.g. tests, debugging, &c)